explainer.md

```diff
@@ -68,356 +107,405 @@ If the set of pages being badged corresponds to an [Installed Web Application](h
 ```
 
-## UX treatment
-Badges may appear in any place that the user agent deems appropriate. In general, these places should be obviously related to the pages being badged, so users understand what the status is for. Appropriate places could include:
-- Tab favicons.
-- Bookmark icons.
-- [OS Specific Contexts](#OS-Specific-Contexts) for [Installed Web Applications](https://www.w3.org/TR/appmanifest/#installable-web-applications).
-  
-> Note: When showing an badge in an [OS Specific Context](#OS-Specific-Contexts) user agents should attempt reuse existing [operating system APIs and conventions](#Specific-operating-system-treatment-for-installed-web-applications), to achieve a native look-and-feel.
-
-If the exact representation of a badge is not supported (e.g., a 2-digit number while only single characters are allowed, or a character where only numbers are allowed) the user agent should make a best effort attempt to map the unsupported data onto supported data. This may involve:
-- Saturating a number: 351 -> '99+'
-- Degrading the data: 7 -> 'flag' when only flag badges are supported.
-
-## Specific operating system treatment for installed web applications
+## A case for separation
 
-This section describes a possible treatment on each major OS. User agents are
-free to implement however they like, but this should give an idea of what the
-API will look like in practice.
+Having enumerated all of the complexity relating to "document" versus "handle"
+contexts, I think we can make a case for separating them into two distinct APIs.
+Under this alternative, there would be an API for setting a badge on URL scopes
+that would *only* be displayed on URL handles such as apps and bookmarks, and
+another API for setting a badge on the current document.
 
-### macOS
+This would mean we no longer have a quick solution for setting the badge
+everywhere on the current origin: to do that, you would need to call both APIs,
+the latter on each page load.
 
-Requires macOS 10.5 or higher.
+The reason to do this is that the two APIs have very different concerns which
+would be properly separated:
 
-* The host API is a string, truncated from the center to a system-defined
-  length, as shown/described
-  [here](https://eternalstorms.wordpress.com/2016/10/29/how-to-badge-an-apps-icon-in-the-dock/).
-  The API is
-  [`NSDockTile.setBadgeLabel`](https://developer.apple.com/documentation/appkit/nsdocktile/1524433-badgelabel).
-* Flag badges are passed as an empty string to `setBadgeLabel` (if that works,
-  otherwise pass a space character). This shows an empty red circle.
-* Integer badges are saturated to 3 or 4 digits (with a "+" if overflowing), and
-  passed as a string to `setBadgeLabel`.
+* The URL scoping complexity only applies to the "handle" API.
+* The push messaging complexity only applies to the "handle" API.
+* We wouldn't need `canBadgeDocument`; the presence of the "document" API itself
+  would indicate whether the tab will be badged in the UI.
```

Comment:
Should we make a note that sites that do wish to do both could use some kind of third party library to do this?

The library could even polyfill the document badging, if it isn't available.

If we decide what our API should look like, I'm happy to throw one together.

Reply:
I guess the library would look like our old API and store its state in indexeddb or something so it can retrieve it on load and set it dynamically? Seems messy, since the indexeddb state could get out of sync with the user agent's state.

I think if we split the APIs, the best practices will be for developers to just treat them separately.